src/mlpack/core/optimizers/adagrad/ada_grad.hpp

```diff
@@ -0,0 +1,155 @@
+/**
+ * @file ada_grad.hpp
+ * @author Abhinav Moudgil
+ *
+ * Implementation of the Adagrad optimizer. Adagrad is an optimizer that
+ * chooses learning rate dynamically by adapting to the data. Hence Adagrad
+ * eliminates the need to manually tune the learning rate.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef __MLPACK_CORE_OPTIMIZERS_ADAGRAD_ADA_GRAD_HPP
+#define __MLPACK_CORE_OPTIMIZERS_ADAGRAD_ADA_GRAD_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * Adagrad is a modified version of stochastic gradient descent which performs
+ * larger updates for more sparse parameters and smaller updates for less sparse
+ * parameters.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @article{duchi2011adaptive,
+ *   author    = {Duchi, John and Hazan, Elad and Singer, Yoram},
+ *   title     = {Adaptive subgradient methods for online learning and stochastic optimization},
+ *   journal   = {Journal of Machine Learning Research},
+ *   year      = {2011}
```

Comment:
This is really pedantic... :)  You can add these too the reference also to make it more complete:

```
  volume={12},
  number={Jul},
  pages={2121--2159},
```